File: BOJ/5676.cpp
#include <cstdio>
#include <algorithm>
#include <cstring>
#include <vector>
#include <queue>
#include <map>
#include <iostream>
#include <string>
#include <math.h>
#include <set>
#include <list>
#include <climits>
#include <string.h>
#include <deque>
#include <functional>
#include <stack>
using namespace std;
typedef long long ll;
#define INF 1000000000
typedef pair<int, int> P;
typedef pair<pair<int, int>, int> PP;
typedef pair<int, pair<int, int>> PPP;
int gox[4] = { 0,1,-1,0 };
int goy[4] = { 1,0,0,-1 };
vector<int> seg;
int h = 1;
int n,k;
#define mod 10007
void update(int index, int value) {
	index += h - 1;
	seg[index] = value;
	while (index > 1) {
		index /= 2;
		seg[index] = seg[index * 2] * seg[index * 2 + 1];
		seg[index] %= mod;
	}
}
int query(int nodenum, int left, int right, int nodeleft, int noderight) {
	if (left > noderight || right < nodeleft) return 1;
	else if (left <= nodeleft && noderight <= right) return seg[nodenum];
	int mid = (nodeleft + noderight) / 2;
	return query(nodenum * 2, left, right, nodeleft, mid)%mod*query(nodenum * 2 + 1, left, right, mid + 1, noderight)%mod;
}
int main() {
	while (scanf(" %d %d", &n, &k) != EOF) {
		seg.clear();
		h = 1;
		while (h < n) h *= 2;
		seg.resize(h * 2);
		vector<char> v;
		for (int i = 1; i <= n; i++) {
			int num;
			scanf("%d", &num);
			update(i, num);
		}
		for (int i = 1; i <= k; i++) {
			char c;
			scanf(" %c", &c);
			if (c == 'C') {
				int index, value;
				scanf("%d%d", &index, &value);
				update(index, value);
			}
			else {
				int left, right;
				scanf("%d%d", &left, &right);
				int ans = query(1, left, right, 1, h);
				if (ans > 0) v.push_back('+');
				else if (ans == 0) v.push_back('0');
				else v.push_back('-');
			}
		}
		for (int i = 0; i < v.size(); i++) printf("%c", v[i]);
		printf("\n");
	}
	
	return 0;
}

File: BOJ(Java 공부용)/5676.java
import java.io.*;
import java.util.StringTokenizer;
public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    static int[] gox = {0, 1, -1, 0, 1, 1, -1, -1};
    static int[] goy = {1, 0, 0, -1, 1, -1, 1, -1};

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String s;
        while ((s = br.readLine()) != null) {
            s = s.trim();
            String[] input = s.split(" ");
            int n = Integer.parseInt(input[0]), m = Integer.parseInt(input[1]);
            StringTokenizer st = new StringTokenizer(br.readLine());
            int h = 1;
            while (h < n) h <<= 1;
            int[] seg = new int[h * 2];
            for (int i = 1; i <= n; i++) {
                int num = Integer.parseInt(st.nextToken());
                update(i, num, h, seg);
            }
            StringBuilder sb = new StringBuilder();
            for (int i = 0; i < m; i++) {
                st = new StringTokenizer(br.readLine());
                String c = st.nextToken();
                if (c.equals("C")) {
                    int index = Integer.parseInt(st.nextToken()), value = Integer.parseInt(st.nextToken());
                    update(index, value, h, seg);
                } else {
                    int left = Integer.parseInt(st.nextToken()), right = Integer.parseInt(st.nextToken());
                    int ans = query(1, left, right, 1, h, seg);
                    if (ans == 1) sb.append('+');
                    else if (ans == 0) sb.append('0');
                    else sb.append('-');
                }
            }
            bw.write(sb.toString()+"\n");
        }
        bw.flush();
        bw.close();
    }

    private static void update(int index, int value, int h, int[] seg) {
        index += h - 1;
        if (value > 0) seg[index] = 1;
        else if (value == 0) seg[index] = 0;
        else seg[index] = -1;
        while (index > 1) {
            index /= 2;
            seg[index] = seg[index * 2] * seg[index * 2 + 1];
        }
    }

    private static int query(int nodenum, int left, int right, int nodeleft, int noderight, int[] seg) {
        if (left > noderight || nodeleft > right) return 1;
        else if (left <= nodeleft && noderight <= right) return seg[nodenum];
        int mid = (nodeleft + noderight) / 2;
        return query(nodenum * 2, left, right, nodeleft, mid, seg) * query(nodenum * 2 + 1, left, right, mid + 1, noderight, seg);
    }
}